File: Chess/src/main/java/game/gui/TilePanel.java
package game.gui;

import game.board.Board;
import game.board.Tile;
import game.move.Move;
import game.piece.Coordinate;
import game.piece.PieceTypes;
import game.piece.Team;
import game.resource.BoardConfigurations;
import game.resource.GuiConfigurations;
import game.util.BoardUtilities;
import game.util.MoveUtilities;
import msg.Message;
import msg.MovementMessage;
import side_client.Client;

import javax.swing.*;
import java.awt.*;
import java.awt.event.MouseEvent;
import java.awt.event.MouseListener;


public class TilePanel extends JPanel {

    Coordinate coordinate;
    JLabel pieceIcon;

    public TilePanel(BoardPanel boardPanel, Coordinate coord, Board chessBoard, Client client) {
        super(new GridBagLayout());
        this.coordinate = coord;
        pieceIcon = new JLabel();
        this.add(pieceIcon);
        setPreferredSize(new Dimension(BoardConfigurations.TILE_SIZE, BoardConfigurations.TILE_SIZE));
        assignTileColor(chessBoard);
        assignTilePieceIcon(chessBoard);
        addMouseListener(new MouseListener() {
            @Override
            public void mouseClicked(MouseEvent e) {
                if (client.getTeam() != chessBoard.getCurrentPlayer().getTeam()) {
                    return;
                }

                if (!chessBoard.hasChosenTile()) {
                    if (chessBoard.getTile(coordinate).hasPiece()) {
                        if (chessBoard.getCurrentPlayer().getTeam() != chessBoard.getTile(coordinate).getPiece().getTeam()) {
                            return;
                        }
                    }

                    chessBoard.setChosenTile(chessBoard.getTile(coordinate));

                } else {
                    Tile destinationTile = chessBoard.getTile(coordinate);
                    if (MoveUtilities.isValidMove(chessBoard, destinationTile)) {
                        Move move = new Move(chessBoard, chessBoard.getChosenTile(), destinationTile);
                        chessBoard.getCurrentPlayer().makeMove(chessBoard, move);
                        if (move.hasKilledPiece()) {
                            client.game.getBottomGameMenu().killedPiecesListModel.addElement(move.getKilledPiece().toString());
                        }


                        Message msg = new Message(Message.MessageTypes.MOVE);
                        MovementMessage movement = new MovementMessage();
                        movement.currentCoordinate = move.getCurrentTile().getCoordinate();
                        movement.destinationCoordinate = move.getDestinationTile().getCoordinate();
                        if (move.getKilledPiece() != null) {
                            movement.isPieceKilled = true;
                        }
                        msg.content = (Object) movement;
                        client.send(msg);
                        chessBoard.changeCurrentPlayer();
                        client.game.getBottomGameMenu().getTurnLBL().setText("Enemy Turn");
                        client.game.getBottomGameMenu().getTurnLBL().setForeground(Color.RED);

                        if (move.hasKilledPiece()) {
                            if (move.getKilledPiece().getType() == PieceTypes.KING) {
                                Team winnerTeam;
                                winnerTeam = (move.getKilledPiece().getTeam() == Team.BLACK) ? Team.WHITE : Team.BLACK;
                                JOptionPane.showMessageDialog(null, "Winner: " + winnerTeam.toString());
                                Message message = new Message(Message.MessageTypes.END);
                                message.content = null;
                                client.send(message);
                            }
                        }

                    } else {
                        if (destinationTile.hasPiece()) {
                            if (chessBoard.getCurrentPlayer().getTeam() != chessBoard.getTile(coordinate).getPiece().getTeam()) {
                                return;
                            }
                        }
                        chessBoard.setChosenTile(destinationTile);

                    }
                    if (MoveUtilities.controlCheckState(chessBoard, Team.BLACK)) {
                        JOptionPane.showMessageDialog(null, "Check state for team : " + Team.BLACK.toString());


                        Message msg = new Message(Message.MessageTypes.CHECK);

                        msg.content = (Object) Team.BLACK;
                        client.send(msg);
                    } else if (MoveUtilities.controlCheckState(chessBoard, Team.WHITE)) {
                        JOptionPane.showMessageDialog(null, "Check state for team : " + Team.WHITE.toString());

                        Message msg = new Message(Message.MessageTypes.CHECK);

                        msg.content = (Object) Team.WHITE;
                        client.send(msg);
                    }
                }
                boardPanel.updateBoardGUI(chessBoard);

            }

            @Override
            public void mousePressed(MouseEvent e) {
            }

            @Override
            public void mouseReleased(MouseEvent e) {
            }

            @Override
            public void mouseEntered(MouseEvent e) {
            }

            @Override
            public void mouseExited(MouseEvent e) {
            }
        });
        validate();
    }

    public Coordinate getCoordinate() {
        return coordinate;
    }

    public void setCoordinate(Coordinate coordinate) {
        this.coordinate = coordinate;
    }

    public void assignTilePieceIcon(Board board) {

        Tile thisTile = board.getTile(this.coordinate);
        if (thisTile == null) {
            System.out.println("Tile is null");
            return;

        }
        if (thisTile.hasPiece()) {


            pieceIcon.setIcon(BoardUtilities.getImageOfTeamPiece(thisTile.getPiece().getTeam(), thisTile.getPiece().getType()));
            pieceIcon.validate();
        } else if (!thisTile.hasPiece()) {
            pieceIcon.setIcon(null);
            pieceIcon.validate();
        }


    }

    public void assignTileColor(Board board) {

        if (this.coordinate.getX() % 2 == 0 && this.coordinate.getY() % 2 == 0) {
            this.setBackground(GuiConfigurations.creamColor);
        } else if (this.coordinate.getX() % 2 == 0 && this.coordinate.getY() % 2 == 1) {
            this.setBackground(GuiConfigurations.lightCyanColor);
        } else if (this.coordinate.getX() % 2 == 1 && this.coordinate.getY() % 2 == 0) {
            this.setBackground(GuiConfigurations.lightCyanColor);
        } else if (this.coordinate.getX() % 2 == 1 && this.coordinate.getY() % 2 == 1) {
            this.setBackground(GuiConfigurations.creamColor);
        }
        if (board.hasChosenTile()) {
            if (this.coordinate.equals(board.getChosenTile().getCoordinate())) {
                this.setBackground(Color.GREEN);
            }

        }
        this.setBorder(BorderFactory.createLineBorder(Color.BLACK));

    }
}

File: Chess/src/main/java/game/gui/TilePanel.kt
package game.gui

import game.Data
import game.Move
import game.board.Board
import game.board.TileNull
import game.piece.Coordinate
import game.piece.PieceNull
import game.piece.PieceTypes
import game.piece.Team
import game.util.BoardUtilities.getImageOfTeamPiece
import game.util.MoveUtilities.controlCheckState
import game.util.MoveUtilities.isValidMove
import msg.Message
import msg.MovementMessage
import side_client.Client
import java.awt.Color
import java.awt.Dimension
import java.awt.GridBagLayout
import java.awt.event.MouseEvent
import java.awt.event.MouseListener
import javax.swing.BorderFactory
import javax.swing.JLabel
import javax.swing.JOptionPane
import javax.swing.JPanel

class TilePanel(boardPanel: BoardPanel, var coordinate: Coordinate, chessBoard: Board, client: Client) : JPanel(
    GridBagLayout()
) {
    private var pieceIcon: JLabel = JLabel()

    init {
        this.add(pieceIcon)
        preferredSize = Dimension(Data.TILE_SIZE, Data.TILE_SIZE)
        assignTileColor(chessBoard)
        assignTilePieceIcon(chessBoard)
        addMouseListener(object : MouseListener {
            override fun mouseClicked(e: MouseEvent) {
                if (client.team !== chessBoard.currentPlayer.team) {
                    return
                }
                if (!chessBoard.hasChosenTile()) {
                    if (chessBoard.getTile(coordinate).hasPiece()) {
                        if (chessBoard.currentPlayer.team !== chessBoard.getTile(coordinate).piece.team) {
                            return
                        }
                    }
                    chessBoard.getTile(coordinate).let { chessBoard.setChosenTile2(it) }
                } else {
                    val destinationTile = chessBoard.getTile(coordinate)
                    if (isValidMove(chessBoard, destinationTile)) {
                        val move = Move(chessBoard, chessBoard.chosenTile, destinationTile)
                        chessBoard.currentPlayer.makeMove(chessBoard, move)
                        if (move.hasKilledPiece()) {
                            client.game.bottomGameMenu.killedPiecesListModel.addElement(move.killedPiece.toString())
                        }
                        val msg = Message(Message.MessageTypes.MOVE)
                        val movement = MovementMessage()
                        movement.currentCoordinate = move.currentTile.coordinate
                        movement.destinationCoordinate = move.destinationTile.coordinate
                        if (move.killedPiece !is PieceNull) {
                            movement.isPieceKilled = true
                        }
                        msg.content = movement
                        client.send(msg)
                        chessBoard.changeCurrentPlayer()
                        client.game.bottomGameMenu.turnLBL.text = "Enemy Turn"
                        client.game.bottomGameMenu.turnLBL.foreground = Color.RED
                        if (move.hasKilledPiece()) {
                            if (move.killedPiece.type === PieceTypes.KING) {
                                val winnerTeam: Team = if (move.killedPiece.team === Team.BLACK) Team.WHITE else Team.BLACK
                                JOptionPane.showMessageDialog(null, "Winner: $winnerTeam")
                                val message = Message(Message.MessageTypes.END)
                                message.content = null
                                client.send(message)
                            }
                        }
                    } else {
                        if (destinationTile.hasPiece()) {
                            if (chessBoard.currentPlayer.team !== chessBoard.getTile(coordinate).piece.team) {
                                return
                            }
                        }
                        chessBoard.setChosenTile2(destinationTile)
                    }
                    if (controlCheckState(chessBoard, Team.BLACK)) {
                        JOptionPane.showMessageDialog(null, "Check state for team : " + Team.BLACK)
                        val msg = Message(Message.MessageTypes.CHECK)
                        msg.content = Team.BLACK
                        client.send(msg)
                    } else if (controlCheckState(chessBoard, Team.WHITE)) {
                        JOptionPane.showMessageDialog(null, "Check state for team : " + Team.WHITE)
                        val msg = Message(Message.MessageTypes.CHECK)
                        msg.content = Team.WHITE
                        client.send(msg)
                    }
                }
                boardPanel.updateBoardGUI(chessBoard)
            }

            override fun mousePressed(e: MouseEvent) {}
            override fun mouseReleased(e: MouseEvent) {}
            override fun mouseEntered(e: MouseEvent) {}
            override fun mouseExited(e: MouseEvent) {}
        })
        validate()
    }

    fun assignTilePieceIcon(board: Board) {
        val thisTile = board.getTile(coordinate)
        if (thisTile is TileNull) {
            println("Tile is null")
            return
        }
        if (thisTile.hasPiece()) {
            pieceIcon.icon = getImageOfTeamPiece(thisTile.piece.team, thisTile.piece.type)
            pieceIcon.validate()
        } else if (!thisTile.hasPiece()) {
            pieceIcon.icon = null
            pieceIcon.validate()
        }
    }

    fun assignTileColor(board: Board) {
        if (coordinate.x % 2 == 0 && coordinate.y % 2 == 0) {
            background = Data.creamColor
        } else if (coordinate.x % 2 == 0 && coordinate.y % 2 == 1) {
            background = Data.lightCyanColor
        } else if (coordinate.x % 2 == 1 && coordinate.y % 2 == 0) {
            background = Data.lightCyanColor
        } else if (coordinate.x % 2 == 1 && coordinate.y % 2 == 1) {
            background = Data.creamColor
        }
        if (board.hasChosenTile()) {
            if (coordinate == board.chosenTile.coordinate) {
                background = Color.GREEN
            }
        }
        border = BorderFactory.createLineBorder(Color.BLACK)
    }
}